draft-ietf-jsonpath-base.md

```diff
@@ -1456,19 +1459,19 @@ The type correctness of function expressions can now be defined in terms of this
 
 A function expression is correctly typed if all the following are true:
 
-* If it occurs as a `filter-path` in a test expression, the function
-is defined to have result type `OptionalNodesType` or one of its subtypes,
-or to have result type `OptionalBooleanType` or one of its subtypes.
+* If it occurs directly in a test expression, the function
+is defined to have a result type out of `ST(OptionalNodesType)`.
+or to have a result type out of `ST(OptionalBooleanType)`.
```

Comment:
> or to have a result type out of `ST(OptionalBooleanType)`.

This is the contradiction.  A boolean cannot rightly occur in a test expression.  The grammar and the type system misalign here.

Reply:
I wasn't trying to change the type system here, just to make the grammar more transparent.
Again, the grammar is not defining the type system (these bullet points are trying to), but what can actually be written.  We don't allow literals to be notated in a test-expr, outside of a comparison-expr or the arguments for a function-expo.  That doesn't mean all the values that these literals can have are invalid as a function return that goes into a test-expr.

Reply:
> the grammar is not defining the type system

I'm not saying that one has to define the other.  I'm saying that what they each define needs to align, and right now they don't.

> That doesn't mean all the values that these literals can have are invalid as a function return that goes into a test-expr.

Maybe this isn't coming across right, but I think you're saying that because `match()` can potentially return `Nothing` that it should be allowed to act as `test-expr`.  Is that what you're saying?

I disagree with this (as mentioned in https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/issues/387#issuecomment-1424866809): 

> Because `match()` has the _potential_ to return these JSON values, the function MUST be restricted to locations where those values are valid, namely `comparable`.

Reply:
> I'm saying that what they each define needs to align

I am not aware of a reason for this.
(I'm hearing you loud and clear that you *want* them to align, and I don't agree.)

Since this discussion really is one of the other issues, I'll merge this PR now, and we can continue the discussion there.